manta-trusted-setup/src/groth16/ppot/hashing.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities to match hashing to G2 curve used in Perpetual Powers of Tau Ceremony
+
+use crate::{
+    groth16::{kzg::G1, ppot::kzg::PerpetualPowersOfTauCeremony},
+    ratio::HashToGroup,
+    util::{hash_to_group, BlakeHasher, Serializer},
+};
+use ark_bn254::{Fq, Fq2, G1Affine, G2Affine};
+use blake2::Digest;
+use manta_crypto::{
+    arkworks::{
+        ec::{short_weierstrass_jacobian::GroupAffine, ProjectiveCurve, SWModelParameters},
+        ff::{BigInteger256, Fp256, FpParameters, Zero},
+    },
+    rand::{RngCore, Sample},
+};
+use manta_util::into_array_unchecked;
+
+/// The G2 hasher used in the PPoT ceremony
+pub struct PpotHasher {
+    /// Domain separation for tau, alpha, beta
+    pub domain_tag: u8,
+}
+
+impl<S, const POWERS: usize, const N: usize>
+    HashToGroup<PerpetualPowersOfTauCeremony<S, POWERS>, [u8; N]> for PpotHasher
+where
+    S: Serializer<G1Affine, G1>,
+{
+    #[inline]
+    fn hash(&self, challenge: &[u8; N], ratio: (&G1Affine, &G1Affine)) -> G2Affine {
```

Comment:
I agree, `pair` is a better name. I probably called it `ratio` because I think of that pair as representing a ratio, but it's probably clearer to say `pair`.